File: 214-shortest-palindrome/shortest-palindrome.cpp
#include<algorithm>
using namespace std;

class Solution {
public:
    string shortestPalindrome(string s) {
        string rev = s;
        reverse(rev.begin(), rev.end());

        string temp = s + '#' + rev;

        vector<int> lps(temp.size(), 0);

        for(int i=1; i<temp.size(); i++){
            int len = lps[i-1];

            while(len>0 && temp[i] != temp[len]){
                len = lps[len - 1];
            }

            if(temp[i] == temp[len])
                len++;
            
            lps[i] = len;
        }

        int longest = lps.back();
        string add = s.substr(longest);
        reverse(add.begin(), add.end());
        return add + s;
    }
};

File: 214-shortest-palindrome/shortest-palindrome.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        rev = s[::-1]
        temp = s + "#" + rev

        lps = [0] * len(temp)

        for i in range(1, len(temp)):
            length = lps[i-1]

            while length>0 and temp[i] != temp[length]:
                length = lps[length -1]
            
            if temp[i] == temp[length]:
                length += 1
            
            lps[i] = length
        
        longest = lps[-1]

        add = s[longest:]

        return add[::-1] + s
        # start =0
        # maxLen =1